### ISSUE TITLE
PSA Packet paths and metadata details

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4-spec#443]
[psa-packet-paths.zip](https://github.com/p4lang/p4-spec/files/1388623/psa-packet-paths.zip)

The attached zip file contains:

+ psa-packet-paths-figure.pptx - A figure similar to one Han Wang has created to show the recirculate, resubmit, and clone paths that is already in the document, but it also shows the "normal" packet paths, and uses abbreviations to name them all.  Those abbreviations are described further in the Excel file below.  This is intended to be _all_ possible ways that a packet can enter, leave, or move between ingress / egress, or at least the minimum set of ways supported by PSA compliant systems (implementers are free to extend this set, of course).  If you think PSA should define ways that aren't shown in this figure, it would be good to bring that up ASAP.

+ psa-packet-paths-figure.pdf - The previous figure converted to PDF

+ psa-packet.paths.xlsx - A Microsoft Excel sheet with a proposal for how all metadata fields and packet contents are initialized when the ingress and egress parser start processing a packet, for each of the kinds given an abbreviated name in the figure.  Some of this is mentioned in some places of the current PSA draft spec, but I think a lot of it is not precisely written down anywhere.  I think the PSA spec _should_ make these things as precise -- at least as precise as we can get agreement on.  These are the kinds of things that PSA implementers could easily differ on otherwise, and people writing programs for PSA devices will want to know what behavior they can rely upon.

There are 2 sheets to the Excel document -- you have to click on the sheet names near the bottom of the window to switch between them.  The second sheet "metadata and packet details" is where most of the open questions are, marked in orange, but all of this is open to discussion if any of it can be improved upon.

**Comment by mbudiu-vmw** [comment]
nice

**Comment by vgurevich** [comment]
Hi Andy,

Great stuff and excellent research and systematization as usual!

Here are my thoughts:
1. Ideally, there should be no difference between a CPU port and a regular port, P4-wise, so NFP and NFCPU should be the same, at least in my mind. If a differentiation is needed, it needs to be done in P4 program. This should cover all the orange items on the first sheet.
2. Source for RESUB and CI2E is **both** the ingress packet buffer **and** the ingress deparser. The ingress packet buffer (AKA "the original packet") provides the packet and ingress deparser only provides additional metadata (via the special clone_out or resubmit_out externs).
3. F6 on the second sheet. Same comment as above - the RESUB packet is the **original** (and not ingress-deparsed) packet. Whether or not packet_in will also start with the prepended metadata (or whether it should be represented by a separate resubmit_in extern) is up to @hanw to define (I'd prefer it as a separate extern).
4. Truncation. I am not sure how exactly it works. Probably it should be a function of PRE and if so, it won't be applicable to RESUB for sure and probably RECIRC too. One can argue that truncation can also be done by the output port. If so then it can affect RECIRC packets.
5. D7 on the second sheet. I think that if we add a construct for quick struct initialization (`{}`), we can get rid of explicit "initialized to 0" and with added checks for "used before initialized" it has the potential to make the code safer. 
6. I'd think that the recirculated packet should have ingress_port set to the recirc port. If someone wants to pass the value of the original ingress port, they need to create a header and deparse it at the egress before recirc.

I'll try to provide more feedback, but I want you to have something for now.

Thanks,
Vladimir

**Comment by jafingerhut** [comment]
[psa-packet-paths.zip](https://github.com/p4lang/p4-spec/files/1395146/psa-packet-paths.zip)

Thanks for the feedback.  I have made the following updates based on your feedback above, and attached a new .zip file containing the updated Excel spreadsheet to this comment.

Vladimir comment: Ideally, there should be no difference between a CPU port and a regular port, P4-wise, so NFP and NFCPU should be the same, at least in my mind. If a differentiation is needed, it needs to be done in P4 program. This should cover all the orange items on the first sheet.

Andy response: I've eliminated the orange coloring on the first sheet, and changed the comments saying that perhaps these from/to CPU cases should be merged with the normal from/to port cases, although keeping them separate for now might help remind us to carefully document the control plane interaction for such packets.

Vladimir comment: Source for RESUB and CI2E is both the ingress packet buffer and the ingress deparser. The ingress packet buffer (AKA "the original packet") provides the packet and ingress deparser only provides additional metadata (via the special clone_out or resubmit_out externs).

Andy response: OK.  For CI2E packets, hopefully the contents of cell F23 is a reasonably clear statement of what you described already.  I changed cell F6 contents for RESUB packets to match it, so it now says that RESUB packets have packet_in contents as follows: "same as pre-IngressParser contents of the packet that caused this resubmitted packet to occur (i.e. with NO modifications from IngressParser, Ingress, or ingress Deparser processing)  TBD: Is truncation of payload supported for such packets?"  Note: I know that this distinction isn't shown in the figure.  I considered trying to represent that in the picture, but I think words are a clearer way to convey this than a figure, at least in a figure that tries to show all of these packet flow cases.

Vladimir comment: F6 on the second sheet. Same comment as above - the RESUB packet is the original (and not ingress-deparsed) packet. Whether or not packet_in will also start with the prepended metadata (or whether it should be represented by a separate resubmit_in extern) is up to @hanw to define (I'd prefer it as a separate extern).

Andy response: My response above covers this, I think.  Prepended metadata I haven’t tried to address yet in this sheet, except with TBD question place-holders, until we know more about how we expect to handle this.

Vladimir comment: Truncation. I am not sure how exactly it works. Probably it should be a function of PRE and if so, it won't be applicable to RESUB for sure and probably RECIRC too. One can argue that truncation can also be done by the output port. If so then it can affect RECIRC packets.

Andy response: I have added a note to cell F7 for RESUB packet_in contents: “@vgurevich suggests truncation will not be supported for RESUB packets.”  Cell F7 for RECIRC packet_in contents I have added: “@vgurevich suggests truncation may not be supported for RECIRC packets.”

Vladimir comment: D7 on the second sheet. I think that if we add a construct for quick struct initialization ({}), we can get rid of explicit "initialized to 0" and with added checks for "used before initialized" it has the potential to make the code safer.

Andy response: Does this represent your thinking on metadata initialization?  I have added it as a comment in cell D7 as a reminder: “The PSA architecture does not mandate initialization of user-defined metadata to known values as given as input to the ingress parser.  If a user’s P4 program explicitly initializes all user-defined metadata early on (e.g. in the parser’s start state), then that will flow through the rest of the parser into the ingress control block as one might normally expect.  But this will be left as an option to the user in their P4 PSA programs, not required behavior for all P4 PSA programs.”

Vladimir comment: I'd think that the recirculated packet should have ingress_port set to the recirc port. If someone wants to pass the value of the original ingress port, they need to create a header and deparse it at the egress before recirc.

Andy response: OK, so sounds like no change to the contents of cell G10 describing RECIRC_PORT as the value of ingress_port for RECIRC packet case.  The comment that such a constant should be added to psa.p4 still stands as a reminder.
